src/nbl/video/COpenGL_Swapchain.h

```diff
@@ -0,0 +1,202 @@
+    {
+        if (!images || !images->size())
+            return nullptr;
+        if (images->size() < params.minImageCount)
+            return nullptr;
+        if (images->size() > MaxImages)
+            return nullptr;
+
+        auto extent = (*images)[0]->getCreationParameters().extent;
+        for (auto& img : (*images))
+        {
+            auto& ci = img->getCreationParameters();
+            if (ci.type != asset::IImage::ET_2D)
+                return nullptr;
+            if (ci.arrayLayers != params.arrayLayers)
+                return nullptr;
+            if (ci.mipLevels != 1u)
+                return nullptr;
+            if (ci.extent.width != extent.width)
+                return nullptr;
+            if (ci.extent.height != extent.height)
+                return nullptr;
+        }
+
+        return core::make_smart_refctd_ptr<COpenGL_Swapchain<FunctionTableType>>(std::move(params), _egl, std::move(images), _features, _master, _config, _major, _minor);
+    }
+
+    E_ACQUIRE_IMAGE_RESULT acquireNextImage(uint64_t timeout, IGPUSemaphore* semaphore, IGPUFence* fence, uint32_t* out_imgIx) override
+    {
+        // TODO impl
+        ++m_imgIx;
+        m_imgIx %= static_cast<uint32_t>(m_images->size());
+        out_imgIx[0] = m_imgIx;
+
+        return EAIR_SUCCESS;
+    }
+
+protected:
+    // images will be created in COpenGLLogicalDevice::createSwapchain
+    COpenGL_Swapchain(SCreationParams&& params, const egl::CEGL* _egl, ImagesArrayType&& images, COpenGLFeatureMap* _features, EGLContext _master, EGLConfig _config, EGLint _major, EGLint _minor) :
+        ISwapchain(std::move(params)),
+        m_threadHandler(_egl, static_cast<ISurfaceGL*>(m_params.surface.get())->getInternalObject(), { images->begin(), images->end() }, _features, _master, _config, _major, _minor),
+        m_thread(&CThreadHandler::thread, &m_threadHandler)
+    {
+        m_images = std::move(images);
+    }
+
+    ~COpenGL_Swapchain()
+    {
+        m_threadHandler.terminate(m_thread);
+    }
+
+private:
+    using SThreadHandlerInternalState = FunctionTableType;
+    class CThreadHandler final : public system::IThreadHandler<SThreadHandlerInternalState>
+    {
+    public:
+        CThreadHandler(const egl::CEGL* _egl, EGLNativeWindowType _window, core::SRange<core::smart_refctd_ptr<IGPUImage>> _images, COpenGLFeatureMap* _features, EGLContext _master, EGLConfig _config, EGLint _major, EGLint _minor) :
+            egl(_egl),
+            masterCtx(_master), config(_config),
+            major(_major), minor(_minor),
+            window(_window),
+            thisCtx(EGL_NO_CONTEXT), surface(EGL_NO_SURFACE),
+            features(_features),
+            images(_images)
+        {
+            assert(images.size() <= MaxImages);
+        }
+
+        void requestBlit(uint32_t _imgIx)
+        {
+            auto raii_handler = createRAIIDispatchHandler();
+
+            needToBlit = true;
+            imgIx = _imgIx;
+        }
```

Comment:
this sort of thing will result in a MAILBOX or FIFO_RELAXED presentation (only your last request will be blitted)

Reply:
#145